test/http_test.c

```diff
@@ -0,0 +1,175 @@
+        return BIO_printf(out, "HTTP/1.1 301 Moved Permanently\r\n"
+                          "Location: /%s\r\n\r\n", rpath) > 0; /* same server */
+    }
+    if (BIO_printf(out, "HTTP/1.1 200 OK\r\n") <= 0)
+        return 0;
+    if (is_get) { /* construct new header and body */
+        if ((len = ASN1_item_i2d(rsp, NULL, it)) <= 0)
+            return 0;
+        if (BIO_printf(out, "Content-Type: application/x-x509-ca-cert\r\n"
+                       "Content-Length: %d\r\n\r\n", len) <= 0)
+            return 0;
+        return ASN1_item_i2d_bio(it, out, rsp);
+    } else {
+        return BIO_write(out, hdr, count) == count; /* echo header and body */
+    }
+}
+
+static long http_bio_cb_ex(BIO *bio, int oper, const char *argp, size_t len,
+                           int cmd, long argl, int ret, size_t *processed)
+{
+
+    if (oper == (BIO_CB_CTRL | BIO_CB_RETURN) && cmd == BIO_CTRL_FLUSH)
+        ret = mock_http_server(bio, (BIO *)BIO_get_callback_arg(bio),
+                               (ASN1_VALUE *)x509, x509_it);
+    return ret;
+}
+
+static int test_http_x509(int do_get)
+{
+    X509 *rcert = NULL;
+    BIO *wbio = BIO_new(BIO_s_mem());
+    BIO *rbio = BIO_new(BIO_s_mem());
+    STACK_OF(CONF_VALUE) *headers = NULL;
+    int res = 0;
+
+    if (wbio == NULL || rbio == NULL)
+        goto err;
+    BIO_set_callback_ex(wbio, http_bio_cb_ex);
+    BIO_set_callback_arg(wbio, (char *)rbio);
+
+    rpath = RPATH;
+    rcert = (X509 *)
+        (do_get ?
+         OSSL_HTTP_get_asn1("http://"SERVER":"PORT"/"RPATH,
+                            NULL /* proxy */, NULL /* proxy_port */,
+                            wbio, rbio, NULL /* bio_update_fn */, NULL,
+                            headers, 0 /* maxline */,
+                            0 /* max_resp_len */, 0 /* timeout */,
+                            "application/x-x509-ca-cert", x509_it)
+         :
+         OSSL_HTTP_post_asn1(SERVER, PORT, RPATH, 0 /* use_ssl */,
+                             NULL /* proxy */, NULL /* proxy_port */,
+                             wbio, rbio, NULL /* bio_update_fn */, NULL,
+                             headers, "application/x-x509-ca-cert",
+                             (ASN1_VALUE *)x509, x509_it, 0 /* maxline */,
+                             0 /* max_resp_len */, 0 /* timeout */,
+                             "application/x-x509-ca-cert", x509_it)
+         );
+    res = TEST_ptr(rcert) && TEST_int_eq(X509_cmp(x509, rcert), 0);
+
+ err:
+    X509_free(rcert);
+    BIO_free(wbio);
+    BIO_free(rbio);
+    sk_CONF_VALUE_pop_free(headers, X509V3_conf_free);
+    return res;
+}
+
+static int test_http_get_x509(void)
+{
+    return test_http_x509(1);
+}
+
+static int test_http_post_x509(void)
+{
+    return test_http_x509(0);
+}
+
+void cleanup_tests(void)
+{
+    X509_free(x509);
+}
+
+int setup_tests(void)
+{
+    x509_it = ASN1_ITEM_rptr(X509);
+    if (!TEST_ptr((x509 = load_pem_cert(test_get_argument(0)))))
```

Comment:
Please add this prior to the `test_get_argument(0)` call (see 8d242823ed):

````
    if (!test_skip_common_options()) {
        TEST_error("Error parsing test options\n");
        return 0;
    }
````

Reply:
Done